qt-bugs@ issue : 45332
bugs.kde.org number : none
applied: no
author: Lubos Lunak <l.lunak@kde.org>

Hello,
 
please apply the attached Qt patch. It avoids problems with the backwards 
compatibility hack in XCursors (hash symlinks in the cursors theme). See 
http://freedesktop.org/pipermail/xlibs/2004-March/000249.html for further 
details (no reply, but I didn't really expect one anyway).
  
--- src/kernel/qcursor_x11.cpp
+++ src/kernel/qcursor_x11.cpp
@@ -252,6 +252,8 @@
     data->bg.red   = 255 << 8;
     data->bg.green = 255 << 8;
     data->bg.blue  = 255 << 8;
+    update(); // Xcursor's backward compatibility hack needs the cursor to be created
+              // right after the bitmaps are created and filled with data
 }
 
 
--- src/kernel/qdnd_x11.cpp
+++ src/kernel/qdnd_x11.cpp
@@ -1073,7 +1073,7 @@
 }
 
 
-void QDragManager::updateCursor()
+void QDragManager::createCursors()
 {
     if ( !noDropCursor ) {
 	noDropCursor = new QCursor( ForbiddenCursor );
@@ -1084,7 +1084,10 @@
 	if ( !pm_cursor[2].isNull() )
 	    linkCursor = new QCursor(pm_cursor[2], 0,0);
     }
+}
 
+void QDragManager::updateCursor()
+{
     QCursor *c;
     if ( willDrop ) {
 	if ( global_accepted_action == QDropEvent::Copy ) {
--- src/kernel/qdragobject.cpp
+++ src/kernel/qdragobject.cpp
@@ -268,6 +268,9 @@
     pm_cursor[0] = QPixmap((const char **)move_xpm);
     pm_cursor[1] = QPixmap((const char **)copy_xpm);
     pm_cursor[2] = QPixmap((const char **)link_xpm);
+#if defined(Q_WS_X11)
+    createCursors(); // Xcursors cache can hold only 8 bitmaps (4 cursors)
+#endif
     object = 0;
     dragSource = 0;
     dropWidget = 0;
--- src/kernel/qdragobject.h
+++ src/kernel/qdragobject.h
@@ -251,6 +251,9 @@
     QDragObject * object;
     bool updateMode( ButtonState newstate );
     void updateCursor();
+#if defined(Q_WS_X11)
+    void createCursors();
+#endif
 
     QWidget * dragSource;
     QWidget * dropWidget;
